File: clients/go/model/bullet.go
package model

import (
	"io"

	mStream "../stream"
)

//Bullet -- type for model bullets of weapons
type Bullet struct {
	WeaponType      WeaponType
	UnitId          int32
	PlayerId        int32
	Position        *Vec2Float64
	Velocity        *Vec2Float64
	Damage          int32
	Size            float64
	ExplosionParams *ExplosionParams
}

//NewBullet -- return new bullet of weapons for shooting
func NewBullet(weaponType WeaponType,
	unitID int32, playerID int32, position *Vec2Float64, velocity *Vec2Float64,
	damage int32, size float64, explosionParams *ExplosionParams) *Bullet {
	return &Bullet{
		WeaponType:      weaponType,
		UnitId:          unitID,
		PlayerId:        playerID,
		Position:        position,
		Velocity:        velocity,
		Damage:          damage,
		Size:            size,
		ExplosionParams: explosionParams,
	}
}

//ReadBullet -- get bullet from net stream from LocalRunner
func ReadBullet(reader io.Reader) *Bullet {
	result := &Bullet{
		WeaponType: ReadWeaponType(reader),
		UnitId:     mStream.ReadInt32(reader),
		PlayerId:   mStream.ReadInt32(reader),
		Position:   ReadVec2Float64(reader),
		Velocity:   ReadVec2Float64(reader),
		Damage:     mStream.ReadInt32(reader),
		Size:       mStream.ReadFloat64(reader),
	}

	if mStream.ReadBool(reader) {
		ExplosionParamsValue := ReadExplosionParams(reader)
		result.ExplosionParams = ExplosionParamsValue
	} else {
		result.ExplosionParams = nil
	}
	return result
}

//Write -- write bullet to net stream to LocalRunner
func (value *Bullet) Write(writer io.Writer) {
	mStream.WriteInt32(writer, int32(value.WeaponType))
	mStream.WriteInt32(writer, value.UnitId)
	mStream.WriteInt32(writer, value.PlayerId)
	value.Position.Write(writer)
	value.Velocity.Write(writer)
	mStream.WriteInt32(writer, value.Damage)
	mStream.WriteFloat64(writer, value.Size)
	if value.ExplosionParams == nil {
		mStream.WriteBool(writer, false)
	} else {
		mStream.WriteBool(writer, true)
		(*value.ExplosionParams).Write(writer)
	}
}

File: clients/javascript/model/bullet.js
const Vec2Double = require('./vec2-double').Vec2Double;
const ExplosionParams = require('./explosion-params').ExplosionParams;
class Bullet {
    constructor(weaponType, unitId, playerId, position, velocity, damage, size, explosionParams) {
        this.weaponType = weaponType;
        this.unitId = unitId;
        this.playerId = playerId;
        this.position = position;
        this.velocity = velocity;
        this.damage = damage;
        this.size = size;
        this.explosionParams = explosionParams;
    }
    static async readFrom(stream) {
        let weaponType;
        weaponType = await stream.readInt();
        let unitId;
        unitId = await stream.readInt();
        let playerId;
        playerId = await stream.readInt();
        let position;
        position = await Vec2Double.readFrom(stream);
        let velocity;
        velocity = await Vec2Double.readFrom(stream);
        let damage;
        damage = await stream.readInt();
        let size;
        size = await stream.readDouble();
        let explosionParams;
        if (await stream.readBool()) {
            explosionParams = await ExplosionParams.readFrom(stream);
        } else {
            explosionParams = null;
        }
        return new Bullet(weaponType, unitId, playerId, position, velocity, damage, size, explosionParams);
    }
    async writeTo(stream) {
        let weaponType = this.weaponType;
        await stream.writeInt(weaponType);
        let unitId = this.unitId;
        await stream.writeInt(unitId);
        let playerId = this.playerId;
        await stream.writeInt(playerId);
        let position = this.position;
        await position.writeTo(stream);
        let velocity = this.velocity;
        await velocity.writeTo(stream);
        let damage = this.damage;
        await stream.writeInt(damage);
        let size = this.size;
        await stream.writeDouble(size);
        let explosionParams = this.explosionParams;
        if (explosionParams === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await explosionParams.writeTo(stream);
        }
    }
}
module.exports = { Bullet: Bullet }
